release-22.1: kvserver: don't transfer leases to draining nodes during scatters

aayushshah15 opened the pull request:
Backport 4/4 commits from #79295.

Backport 1/1 commit from #79581.

Backport 1/1 commit from #81403.

/cc @cockroachdb/release

---

**kvserver: introduce Allocator.ValidLeaseTargets()**

This commit is a minor refactor of the `Allocator.TransferLeaseTarget` logic in
order to make it more readable and, to abstract out a new exported `Allocator`
method called `ValidLeaseTargets()`.

The contract of `ValidLeaseTargets()` is as follows:
```
// ValidLeaseTargets returns a set of candidate stores that are suitable to be
// transferred a lease for the given range.
//
// - It excludes stores that are dead, or marked draining or suspect.
// - If the range has lease_preferences, and there are any non-draining,
// non-suspect nodes that match those preferences, it excludes stores that don't
// match those preferences.
// - It excludes replicas that may need snapshots. If replica calling this
// method is not the Raft leader (meaning that it doesn't know whether follower
// replicas need a snapshot or not), produces no results.

```

Previously, there were multiple places where we were performing the logic
that's encapsulated by `ValidLeaseTargets()`, which was a potential source of
bugs. This is an attempt to unify this logic in one place that's relatively
well-tested. This commit is only a refactor, and does not attempt to change any
behavior. As such, no existing tests have been changed, with the exception of a
subtest inside `TestAllocatorTransferLeaseTargetDraining`. See the comment over
that subtest to understand why the behavior change made by this patch is
desirable.

The next commit in this PR uses this method to fix (at least part of) https://github.com/cockroachdb/cockroach/issues/74691.

Release note: none

**kvserver: don't transfer leases to draining nodes during scatters**

Previously, `AdminScatter` called with the `RandomizeLeases` option could
potentially transfer leases to nodes marked draining.  This commit leverages
the refactor from the last commit to fix this bug by first filtering the set of
candidates down to a set of valid candidates that meet lease preferences and
are not marked suspect or draining.

Relates to and fixes a part of https://github.com/cockroachdb/cockroach/issues/74691.

Release note (bug fix): Fixes a bug where draining / drained nodes could
re-acquire leases during an import or an index backfill.

Release justification: bug fix.

aayushshah15 committed:
kvserver: clarify a call into `Allocator.TransferLeaseTarget`

Previously, it wasn't very clear how a call into `maybeTransferLeaseAway` would
find a lease transfer target. This commit makes a small effort to make this
obvious.

Release note: None

aayushshah15 committed:
kvserver: introduce `Allocator.ValidLeaseTargets()`

This commit is a minor refactor of the `Allocator.TransferLeaseTarget` logic in
order to make it more readable and, to abstract out a new exported `Allocator`
method called `ValidLeaseTargets()`.

The contract of `ValidLeaseTargets()` is as follows:
```
// ValidLeaseTargets returns a set of candidate stores that are suitable to be
// transferred a lease for the given range.
//
// - It excludes stores that are dead, or marked draining or suspect.
// - If the range has lease_preferences, and there are any non-draining,
// non-suspect nodes that match those preferences, it excludes stores that don't
// match those preferences.
// - It excludes replicas that may need snapshots. If replica calling this
// method is not the Raft leader (meaning that it doesn't know whether follower
// replicas need a snapshot or not), produces no results.

```

Previously, there were multiple places where we were performing the logic
that's encapsulated by `ValidLeaseTargets()`, which was a potential source of
bugs. This is an attempt to unify this logic in one place that's relatively
well-tested. This commit is only a refactor, and does not attempt to change any
behavior. As such, no existing tests have been changed, with the exception of a
subtest inside `TestAllocatorTransferLeaseTargetDraining`. See the comment over
that subtest to understand why the behavior change made by this patch is
desirable.

The next commit in this PR uses this method to fix (at least part of) #74691.

Release note: none

aayushshah15 committed:
kvserver: don't transfer leases to draining nodes during scatters

Previously, `AdminScatter` called with the `RandomizeLeases` option could
potentially transfer leases to nodes marked draining.  This commit leverages
the refactor from the last commit to fix this bug by first filtering the set of
candidates down to a set of valid candidates that meet lease preferences and
are not marked suspect or draining.

Release note (bug fix): Fixes a bug where draining / drained nodes could
re-acquire leases during an import or an index backfill.

aayushshah15 committed:
kvserver: rename `transferLeaseOptions.checkTransferLeaseSource`

This commit inverts this boolean and renames it to `excludeLeaseRepl`, which is
a lot more intuitive for the callers of `Allocator.TransferLeaseTarget`.

Release note: none

nvb committed:
kv: fix logging of leaseholder's StoreID in ShouldTransferLease

Fixes #79566.

Fallout from a6a8d5c.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/80834)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I did not do a pass on some parts of this before they landed on master, so I left a few comments below. Feel free to address them in a separate PR on master and then add a commit to this backport PR.

Also, we saw in https://github.com/cockroachlabs/support/issues/1569 that this fixes a bug where the allocator could skip the `excludeReplicasInNeedOfSnapshots` protection when lease preferences are present. This backport will fix the bug for release-22.1. We'd like to also fix it for release-21.2. Do you have a sense of the difficulty of a backport of this change to release-21.2? We could do something more targeted if it's too much of a lift.

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, 6 of 6 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-:-N1y_kq31GWZuPDNhdnG:bx0qqqr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15)

___
*[`pkg/kv/kvserver/allocator.go` line 1501 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1yYIr89CH3htwYJ73F:-N1yYIr89CH3htwYJ73G:bupyqlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/62d62fda1907668c71f7e90e9ead7fb19ac5139b/pkg/kv/kvserver/allocator.go#L1501)):*
> ```Go
> 	candidates := make([]roachpb.ReplicaDescriptor, 0, len(existing))
> 	for i := range existing {
> 		if existing[i].GetType() != roachpb.VOTER_FULL {
> ```

In https://github.com/cockroachdb/cockroach/pull/74546, we began allowing `VOTER_INCOMING` replicas to receive the lease. Should we be doing the same here?

Should we be using `CheckCanReceiveLease` in this function?
___
*[`pkg/kv/kvserver/allocator.go` line 1579 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1yYpuC8fYD3acbBr-N:-N1yYpuC8fYD3acbBr-O:bcvzuyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/62d62fda1907668c71f7e90e9ead7fb19ac5139b/pkg/kv/kvserver/allocator.go#L1579)):*
> ```Go
> 	// If there are any replicas that do match lease preferences, then we check if
> 	// the existing leaseholder is one of them.
> 	preferred := a.preferredLeaseholders(conf, candidates)
> ```

Should these preferred replicas be passed through `excludeReplicasInNeedOfSnapshots`, or is the idea that a later call to `ValidLeaseTargets` will perform this check?
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1052 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1yXts57eIaRNyaR3t1:-N1yXts57eIaRNyaR3t2:b-kv02s7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7369ff2f78d6ac8a19acfaca05c93f185576dcb0/pkg/kv/kvserver/replicate_queue.go#L1052)):*
> ```Go
> 		conf,
> 		transferLeaseOptions{
> 			goal:   leaseCountConvergence,
> ```

Is this changing behavior? Isn't `followTheWorkload` the zero-value for this enum? Is the change in behavior intentional?
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 489 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1y_7tQAn5s2FdUWDQm:-N1y_7tQAn5s2FdUWDQn:bea84nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5de16059a38b5fd593e6f7f68c9b6d2372a2a66/pkg/kv/kvserver/store_rebalancer.go#L489)):*
> ```Go
> 			replWithStats.repl.leaseholderStats,
> 			true, /* forceDecisionWithoutStats */
> 			transferLeaseOptions{
> ```

nit: consider adding the `ExcludeLeaseRepl` field back in so that this call is explicit.
___
*[`pkg/kv/kvserver/allocator_test.go` line 1746 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1y_LtB6oUJ4SvqmPli:-N1y_LtCEvle8lLR0sjR:bk18y7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5de16059a38b5fd593e6f7f68c9b6d2372a2a66/pkg/kv/kvserver/allocator_test.go#L1746)):*
> ```Go
> 		existing       []roachpb.ReplicaDescriptor
> 		leaseholder    roachpb.StoreID
> 		allowLeaseRepl bool
> ```

nit: should we use `excludeLeaseRepl` here as well so that readers of the test don't need to negate the input? Same question below.


<!-- Sent from Reviewable.io -->

aayushshah15 committed:
kvserver: address assorted comments from previous refactor of `TransferLeaseTarget()`

This commit addresses miscelleneous review comments from #80834.

Release note: None

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80834)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/allocator.go` line 1501 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1yYIr89CH3htwYJ73F:-N2IPUm_Bas_ZbbQEfGF:b-o7t7pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/62d62fda1907668c71f7e90e9ead7fb19ac5139b/pkg/kv/kvserver/allocator.go#L1501)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In https://github.com/cockroachdb/cockroach/pull/74546, we began allowing `VOTER_INCOMING` replicas to receive the lease. Should we be doing the same here?

Should we be using `CheckCanReceiveLease` in this function?
</blockquote></details>

I changed it to using the `IsVoterNewConfig()` method that `CheckCanReceiveLease` calls into, for now, since plumbing the range descriptor through into `TransferLeaseTarget()` for a call into `CheckCanReceiveLease` has quite a large footprint in terms of the tests that need to be changed. LMK if you prefer something else.
___
*[`pkg/kv/kvserver/allocator.go` line 1579 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1yYpuC8fYD3acbBr-N:-N2IQkn4ADSiV9JyD1Nm:buwj1lv) ([raw file](https://github.com/cockroachdb/cockroach/blob/62d62fda1907668c71f7e90e9ead7fb19ac5139b/pkg/kv/kvserver/allocator.go#L1579)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should these preferred replicas be passed through `excludeReplicasInNeedOfSnapshots`, or is the idea that a later call to `ValidLeaseTargets` will perform this check?
</blockquote></details>

It does seem like a good idea for `leaseholderShouldMoveDueToPreferences()` to not fire if all preferred replicas need snapshots. I've made that change, and looks like nothing breaks (because, as you said, `ValidLeaseTargets()` was filtering these replicas out anyway).
___
*[`pkg/kv/kvserver/replicate_queue.go` line 1052 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1yXts57eIaRNyaR3t1:-N2IRQs8-hIY792iimuu:b-d1msvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7369ff2f78d6ac8a19acfaca05c93f185576dcb0/pkg/kv/kvserver/replicate_queue.go#L1052)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this changing behavior? Isn't `followTheWorkload` the zero-value for this enum? Is the change in behavior intentional?
</blockquote></details>

The thing is that whenever `excludeLeaseRepl` is true, `followTheWorkload` always falls back to `leaseCountConvergence`. I just wanted to make that explicit here for this caller.

It's sort of gross how `TransferLeaseTarget` has become a bit of a monstrosity to use due to all these options subtly interacting with each other. I didn't really attempt to improve that as part of the original PR though. 
___
*[`pkg/kv/kvserver/store_rebalancer.go` line 489 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1y_7tQAn5s2FdUWDQm:-N2IT0_o-65qn2-fIdId:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5de16059a38b5fd593e6f7f68c9b6d2372a2a66/pkg/kv/kvserver/store_rebalancer.go#L489)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider adding the `ExcludeLeaseRepl` field back in so that this call is explicit.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/allocator_test.go` line 1746 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-N1y_LtB6oUJ4SvqmPli:-N2IX25h0oOm7ywZvuJw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a5de16059a38b5fd593e6f7f68c9b6d2372a2a66/pkg/kv/kvserver/allocator_test.go#L1746)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: should we use `excludeLeaseRepl` here as well so that readers of the test don't need to negate the input? Same question below.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
I've opened https://github.com/cockroachdb/cockroach/pull/81403 to address the review comments you left, PTAL. I'll add that commit to this backport PR once you approve it. 

I tried backporting to 21.2 but there have been a bunch of changes in the interim that make it complicated. Even after resolving the conflicts, I don't have a ton of confidence there and we should probably just do a targeted fix for 21.2. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80834#-:-N2IX4pLDe9h4iMAa8z3:b-auu8fx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@aayushshah15 it looks like https://github.com/cockroachdb/cockroach/pull/79295 landed. Do we want to patch that in here and then get this merged?
